### ISSUE TITLE
Internal Database Error with PEP and Postgresql 9.5.

### ISSUE THREAD

**Issue by irmlerjo** [processone/ejabberd#2014]
> What version of ejabberd are you using?
17.08.50

> What operating system (version) are you using?
Ubuntu 16.04

> How did you install ejabberd (source, package, distribution)?
Installed from Source

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

I am trying to write a Omemo extension using stanza.io and ejabberd, with mod_pubsub and pep, mapped to postgres 9.5 as sql database.
When initially setting a pep node it works fine, but when trying to retrieve that node an internal database error occurs in ejabberd:

`[error] transaction return internal error: {aborted,{badarg,[{erlang,integer_to_binary,[undefined],[]},{node_flat_sql,'-get_items/3-fun-3-',8,[{file,"src/node_flat_sql.erl"},{line,714}]},{ejabberd_sql,sql_query_internal,1,[{file,"src/ejabberd_sql.erl"},{line,580}]},{ejabberd_sql,sql_query_t,1,[{file,"src/ejabberd_sql.erl"},{line,178}]},{node_flat_sql,get_items,3,[{file,"src/node_flat_sql.erl"},{line,716}]},{mod_pubsub,node_call,4,[{file,"src/mod_pubsub.erl"},{line,3526}]},{mod_pubsub,'-transaction/4-fun-0-',3,[{file,"src/mod_pubsub.erl"},{line,3554}]},{ejabberd_sql,execute_bloc,1,[{file,"src/ejabberd_sql.erl"},{line,511}]}]}}`

I tried to look into this, and I did a workaround, by replacing the **Max** variable in node_flat_sql.erl /line 714, by just any integer. After doing so, the result set i got was just as expected. Since I don't know any erlang, i could not find out where this variable is set(did not see it in this file), so this might just be some wrong configuration on my part.

My ejabberd config is:

```
loglevel: 4
log_rotate_size: 10485760
log_rotate_date: ""
log_rotate_count: 1
log_rate_limit: 100
hosts:
  - "localhost"
  - "omemotest"
listen: 
  - 
    port: 5222
    ip: "::"
    module: ejabberd_c2s
    max_stanza_size: 65536
    shaper: c2s_shaper
    access: c2s
  - 
    port: 5269
    ip: "::"
    module: ejabberd_s2s_in
  - 
    port: 5280
    ip: "::"
    module: ejabberd_http
    request_handlers:
      "/ws": ejabberd_http_ws
      "/bosh": mod_bosh
      "/api": mod_http_api
    web_admin: true
    http_bind: true
    captcha: false
    tls: true
    certfile: "/usr/local/etc/ejabberd/ejabberd.pem"
 
host_config:
 "omemotest":
  sql_type: pgsql
  sql_server: "localhost"
  sql_database: "xmpp"
  sql_username: "<<myuser>>"
  sql_password: "<<mypassword>>"
  auth_method: sql
  sql_port: 5433

shaper:

  normal: 1000

  fast: 50000

max_fsm_queue: 1000

acl:

  admin:
    user:
      - "admin" : "omemotest"
  local: 
    user_regexp: ""
  loopback:
    ip:
      - "127.0.0.0/8"
      - "::1/128"
      - "::FFFF:127.0.0.1/128"

shaper_rules:
  max_user_sessions: 10
  max_user_offline_messages:
    - 5000: admin
    - 100
  c2s_shaper:
    - none: admin
    - normal
  s2s_shaper: fast

access_rules:
  local:
    - allow: local
  c2s:
    - deny: blocked
    - allow
  announce:
    - allow: admin
  configure: 
    - allow: admin
  muc_create: 
    - allow: local
  pubsub_createnode: 
    - allow: local
  register: 
    - allow
  trusted_network: 
    - allow: loopback

api_permissions:
  "console commands":
    from:
      - ejabberd_ctl
    who: all
    what: "*"
  "admin access":
    who:
      - access:
          - allow:
            - acl: loopback
            - acl: admin
      - oauth:
        - scope: "ejabberd:admin"
        - access:
          - allow:
            - acl: loopback
            - acl: admin
    what:
      - "*"
      - "!stop"
      - "!start"
  "public commands":
    who:
      - ip: "127.0.0.1/8"
    what:
      - "status"
      - "connected_users_number"
language: "en"
modules: 
  mod_adhoc: {}
  mod_admin_extra: {}
  mod_announce: # recommends mod_adhoc
    access: announce
  mod_blocking: {} # requires mod_privacy
  mod_caps: 
     db_type: sql
  mod_carboncopy: {}
  mod_client_state: {}
  mod_configure: {} # requires mod_adhoc
  mod_disco: {}
  mod_echo: {}
  mod_irc: {}
  mod_bosh: {}
  mod_muc: 
    access:
      - allow
    access_admin:
      - allow: admin
    access_create: muc_create
    access_persistent: muc_create
  mod_muc_admin: {}
  mod_offline: 
    access_max_user_messages: max_user_offline_messages
  mod_ping: {}
  mod_privacy: {}
  mod_private: {}
  mod_pubsub: 
    db_type: sql
    access_createnode: pubsub_createnode
    ignore_pep_from_offline: true
    ## ignore_pep_from_offline: false
    last_item_cache: false
    plugins: 
      - "pep" ## pep requires mod_caps
  mod_push: {}
  mod_push_keepalive: {}
  mod_roster: {}
  mod_shared_roster: {}
  mod_stats: {}
  mod_time: {}
  mod_vcard:
    search: false
  mod_version: {}
  mod_stream_mgmt: {}
  mod_s2s_dialback: {}
  mod_http_api: {}

allow_contrib_modules: true
```

As I'm not a seasoned xmpp nor ejabberd user, I give you a little more information in case I did some trivial mistake in configuring the ejabberdserver

The Stanza used to retrieve that node was:
`<iq xmlns="jabber:client" type="get" to="admin@omemotest" id="ab8038d8-af03-4cd1-b45d-912ef0cd93f2"><pubsub xmlns="http://jabber.org/protocol/pubsub"><items node="urn:xmpp:omemo:0:devicelist"/><set xmlns="http://jabber.org/protocol/rsm"/></pubsub></iq>`

And initially I was sending a stanza like this:
`<iq xmlns="jabber:client" type="set" to="admin@omemotest" id="0bdd8415-c2ff-4b09-b43c-db3be853ff82"><pubsub xmlns="http://jabber.org/protocol/pubsub"><publish node="urn:xmpp:omemo:0:devicelist"><item id="5E2561BBB5E74"><list xmlns="urn:xmpp:omemo:0"><device id="29cc0c2130045bb65d8bf3df0da616f9f5b620fed995b4d00ab9fb8ff835ab55"/></list></item></publish></pubsub></iq>`

Thank you for your time

**Comment by cromain** [comment]
i'm checking RSM now as part of #1994, so i'll check about this as well.

**Comment by irmlerjo** [comment]
My problem is solved when adding the max option in client.
So the only problem was the empty RSM tag, which when flying over the spec does not make any sense anyways it seems.
So if at all this would be a stanza.io issue i think.
Thank you still for the fast response.

**Comment by cromain** [comment]
should be fixed by 07a193d
